torch/distributed/_sharded_tensor/ops/linear.py

```diff
@@ -0,0 +1,128 @@
+import torch
+import torch.distributed as dist
+from torch.distributed._sharding_spec import ChunkShardingSpec
+from torch.distributed._sharding_spec._internals import (
+    get_split_size,
+    get_chunked_dim_size,
+)
+
+
+def sharded_linear(types, args, kwargs, pg):
+    from torch.distributed._sharded_tensor import ShardedTensor
+
+    input = args[0]
+    weight = args[1]
+    bias = kwargs["bias"]
+
+    # Validate types
+    if not isinstance(input, torch.Tensor) or not isinstance(bias, torch.Tensor):
+        raise TypeError("input and bias need to be torch.Tensor")
+    if not isinstance(weight, ShardedTensor):
+        raise TypeError("weight needs to be ShardedTensor")
+
+    if input.size()[1] != weight.size()[1]:
+        raise ValueError(
+            f'Input dim: {input.size()[1]} does not match '
+            f'appropriate weight dim: {weight.size()[1]}')
+    if not isinstance(weight._sharding_spec, ChunkShardingSpec):
+        raise ValueError("Only ChunkShardingSpec supported for ShardedTensor ops!")
+    if len(weight.local_shards()) != 1:
+        raise ValueError("Only one local shard supported!")
+
+    local_shard = weight.local_shards()[0].tensor
+    local_shard_t = local_shard.t().contiguous()
+    sharding_dim = weight._sharding_spec.dim
+    world_size = dist.get_world_size(pg)
+    rank = dist.get_rank(pg)
+
+    if sharding_dim == 1:
+        return _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bias, pg)
+    elif sharding_dim == 0:
+        return _handle_col_wise_sharding(input, world_size, weight, local_shard_t, bias, pg)
+    else:
+        raise RuntimeError(f'nn.Linear weight sharded on dim {sharding_dim} not supported!')
+
+def _handle_col_wise_sharding(input, world_size, weight, local_shard_t, bias, pg):
+    # allgather the inputs first.
+    gathered_inputs = [torch.zeros_like(input) for _ in range(world_size)]
+    dist.all_gather(gathered_inputs, input, group=pg)
+
+    # matmul all the inputs.
+    results = []
+    for i, inp in enumerate(gathered_inputs):
+        results.append(inp.matmul(local_shard_t).t())
+
+    # Process inputs and outputs for all2all.
+    sharding_dim_size = weight.size()[0]
+    output = torch.empty((sharding_dim_size, input.size(0)), device=input.device)
+    combined_results = torch.cat(results)
+
+    # Compute output splits
+    split_size = get_split_size(sharding_dim_size, world_size)
+    output_split_sizes = [
+        get_chunked_dim_size(sharding_dim_size, split_size, placement.rank())
+        for placement in weight._sharding_spec.placements
+    ]
+
+    # distribute the outputs using all2all.
+    dist.all_to_all_single(output, combined_results, output_split_sizes=output_split_sizes, group=pg)
```

Comment:
@cbalioglu The collectives in this file are examples of collectives for which we need backward.